x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -75,12 +81,27 @@ public TransportEsqlQueryAction(
             bigArrays
         );
         this.settings = settings;
+        this.transportService = transportService;
     }
 
     @Override
     protected void doExecute(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
         // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
-        requestExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
+        final ActionListener<EsqlQueryResponse> wrappedListener;
+        final TimeValue timeout = timeout(request);
+        if (timeout != null) {
+            final ThreadPool threadPool = transportService.getThreadPool();
+            final var executor = threadPool.executor(EsqlPlugin.ESQL_THREAD_POOL_NAME);
+            wrappedListener = ListenerTimeouts.wrapWithTimeout(threadPool, timeout, executor, listener, l -> {
+                logger.debug("cancelling ESQL task {} on timeout", task);
+                final TaskManager taskManager = transportService.getTaskManager();
+                taskManager.cancelTaskAndDescendants((CancellableTask) task, "timeout", false, ActionListener.noop());
+                listener.onFailure(new ElasticsearchTimeoutException("ESQL query timed out after {}", timeout));
```

Comment:
We should complete the supplied listener `l` rather than the outer `listener`, otherwise `listener` may be completed twice (and also to avoid unnecessarily capturing `listener` in this lambda)

Reply:
Oh wait sorry, this API is very confusing. What you've written is right.

Reply:
Why do we even pass in a listener to the `onTimeout` callback? This listener is a no-op in all cases isn't it?

Reply:
Yes, I think the API is confusing. I will look into simplifying it.